### ISSUE TITLE
dart:core Uri breaking on Windows

### ISSUE THREAD

**Issue by daniel-v** [dart-lang/sdk#27837]
Quickly reproducible:

```dart
void main() {
  new Uri.file('some_package_name|test/some_file.dart');
}
```

**Error:**
```
Unhandled exception:
Invalid argument(s): Illegal character in path
#0      _Uri._checkWindowsPathReservedCharacters (dart:core/uri.dart:1632)
#1      _Uri._makeWindowsFileUrl (dart:core/uri.dart:1733)
#2      _Uri._Uri.file (dart:core/uri.dart:1598)
#3      main (file:///C:/Users/Dani/Documents/Projects/intl_testbed/test/uri_test.dart:4:17)
#4      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:261)
#5      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
```

While one would not want to create a Uri from `some_package_name|test/some_file.dart` it is often passed by transformers. One affected package that passes such uri string to the constuctor is analyzer via `intl_translation` package.

Stack trace for that:

```
Transform IntlMessage on some_package_name|web/main.dart threw error: Invalid argument(s): Illegal character in path
dart:core/uri.dart 1632                                   _Uri._checkWindowsPathReservedCharacters
dart:core/uri.dart 1733                                   _Uri._makeWindowsFileUrl
dart:core/uri.dart 1598                                   _Uri._Uri.file
package:analyzer/src/string_source.dart 30:45             StringSource.StringSource
package:analyzer/analyzer.dart 43:23                      parseCompilationUnit
package:intl_translation/src/message_rewriter.dart 42:23  findMessages
package:intl_translation/src/message_rewriter.dart 17:18  rewriteMessages
package:intl_translation/transformer.dart 23:22           IntlMessageTransformer.apply.<async>
```

I consider this *critical* as I keep running into this.

**Comment by zoechi** [comment]
I assumed that this form (`'some_package_name|test/some_file.dart'`)  is just how it's printed in some error messages or similar.

I'd expect it to be new 
```dart
Uri.file('packages/some_package_name/test/some_file.dart');
```

**Comment by daniel-v** [comment]
Unfortunately, the situation is not that simple.

What I got so far debugging is that the `fullName` parameter in this line of the analyzer package looks just like what I wrote.
https://github.com/dart-lang/sdk/blob/master/pkg/analyzer/lib/src/string_source.dart#L30

So the Uri constructor in fact receives a String that follows the format of: `some_package_name|test/some_file.dart`

I tested this on a linux machine as well (if memory serves right), the fullName follows the same pattern. 

Could this Dartpad help?
https://dartpad.dartlang.org/8fb08da0345674a454e2f9559a6d7a84

**Comment by zoechi** [comment]
Have you tried the format I posted?
I may miss something but `'some_package_name|test/some_file.dart'` is not a known URI format, at least not to me.

**Comment by daniel-v** [comment]
No, I have not. The reason why I did not do that is because the Uri constructor receives the path in such a format. It works fine on linux and fails on windows (which is why I feel it painful).

I spent some more time trying to figure things out. Here is some code that can help with the reproduction. Please try it on Linux and Win to see the difference.

https://github.com/daniel-v/transformer_repro
Just do a `pub serve web` to execute the transformer of intl_translation.

**new information**:
it looks like that when modifying the pubspec.yaml to look like this:
```yaml
transformers:
- intl_translation:
    $include: web\main.dart
```
the error's gone, seems to work ok.

**Comment by lrhn** [comment]
The problem is that the string `some_package_name|test/some_file.dart` is not a valid *file name* on Windows. The `Uri.file` constructor with Windows semantics rejects paths with segments that contain any of the characters `"*/:<>?\|`, so this gets rejected for both `/` and `|`.

On non-Windows, only the character `/` is rejected in segments, and since the input path is split on `/`, that can't happen.

You can reproduce this on non-Windows by using the `windows: true` parameter to `Uri.file`. That also suggests a solution: Pass `windows: false` where you call `Uri.file` if you don't want to reject Windows-invalid paths. That won't spit the path on `\` characters, but on the other hand, it will accept `\` and `|` in the path segments.

So, in short: The argument here is not a local file path (it can't be, file paths can't contain those characters) so converting it to a `file:` URI is rejected deliberately.

Uri.file` doesn't just prepend `file:///` to the path, it actually checks that it's a valid file path and on Windows (or in Windows mode) it parses it for drive, network or UNC prefixes. Use `windows: false` to not do that.

**Comment by daniel-v** [comment]
@lrhn Thank you for the lengthy explanation, I appreciate you putting effort into giving details why it is labelled as "works as intended". I feel this is a good practice.

Even though I still can't use the package mentioned, it seems clearer that it is not a failure on the dart:core's part.

**Comment by mso7nueh** [comment]
![image](https://user-images.githubusercontent.com/90323051/217037021-d0ffe289-fa7b-4efa-9a57-d15692338b6d.png)
I am trying to run a file from assets on the windows platform, but I get the following error. Please help me fix this.
